modules/living_spaces_event/src/Controller/LivingSpacesEventIcalController.php

```diff
@@ -0,0 +1,151 @@
+
+  /**
+   * Returns the file.repository service.
+   *
+   * @var \Drupal\file\FileRepositoryInterface
+   */
+  protected $fileRepository;
+
+  /**
+   * Constructs a new LivingSpacesEventIcalController object.
+   *
+   * @param \Drupal\Core\Database\Connection $database
+   *   The database connection to be used.
+   * @param \Symfony\Component\HttpFoundation\RequestStack $request
+   *   Forward-compatibility shim for Symfony's RequestStack.
+   * @param \Drupal\Core\File\FileSystemInterface $file_system
+   *   Provides an interface for helpers that operate on files and stream wrappers.
+   * @param \Drupal\file\FileRepositoryInterface $file_repository
+   *   Performs file system operations and updates database records accordingly.
+   */
+  public function __construct(Connection $database, RequestStack $request, FileSystemInterface $file_system, FileRepositoryInterface $file_repository) {
+    $this->database = $database;
+    $this->request = $request;
+    $this->fileSystem = $file_system;
+    $this->fileRepository = $file_repository;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('database'),
+      $container->get('request_stack'),
+      $container->get('file_system'),
+      $container->get('file.repository')
+    );
+  }
+
+  /**
+   * Callback for 'iCal' route.
+   */
+  public function download(GroupInterface $group) {
+    $host = $this->request->getCurrentRequest()->getHost();
+    $calendar = new Calendar($host);
+
+    $query = $this->database->select('living_spaces_event_field_data', 'efd');
+    $query->fields('efd', ['id', 'label', 'description__value', 'location__value']);
+    $query->condition('efd.space', $group->id());
+    $query->innerJoin('living_spaces_event__field_start_date', 'efsd', 'efsd.entity_id = efd.id');
+    $query->addField('efsd', 'field_start_date_value');
+    $query->innerJoin('living_spaces_event__field_end_date', 'efed', 'efed.entity_id = efd.id');
+    $query->addField('efed', 'field_end_date_value');
+
+    if ($results = $query->execute()->fetchAllAssoc('id')) {
+      foreach ($results as $result) {
+        $event = new Event();
+        $event->setDtStart(new \DateTime($result->field_start_date_value, new \DateTimeZone('UTC')));
+        $event->setDtEnd(new \DateTime($result->field_end_date_value, new \DateTimeZone('UTC')));
+        $event->setSummary($result->label);
+        $event->setDescription($result->description__value);
+        $event->setLocation($result->location__value);
+        $event->setUrl( "{$host}/living-spaces-event/{$result->id}");
+        $calendar->addComponent($event);
+      }
+    }
+
+    $filename = "iCal {$group->label()}.ics";
+    $uri = 'public://' . $filename;
+
+    $content = $calendar->render();
+    try {
+      $file = $this->fileRepository->writeData($content, $uri, FileSystemInterface::EXISTS_REPLACE);
+    }
+    catch (\Exception $e) {
```

Comment:
We should at least add some logs to watchdog about this exception.
Also I don't like the idea users don't see any messaging that something went wrong.
In case of error we should let user know that something went wrong and he should try later, while also logging details to watchdog for us to investigate.

Reply:
`writeData()` method already adds logs. So, simply added a message.